src/backend/access/transam/csn_log.c

```diff
@@ -0,0 +1,600 @@
+ * in a database restart, standby should apply wal record for GUC changed,
+ * then it's difficult to notice all backends about that. So they can get
+ * the message by 'csnSnapshotActive' which in share buffer. It will not
+ * acquire a lock, so without performance issue.
+ *
+ */
+typedef struct CSNshapshotSharedData
+{
+	bool		csnSnapshotActive;
+} CSNshapshotSharedData;
+
+typedef CSNshapshotSharedData *CSNshapshotShared;
+CSNshapshotShared csnShared = NULL;
+
+/*
+ * Defines for CSNLog page sizes.  A page is the same BLCKSZ as is used
+ * everywhere else in Postgres.
+ *
+ * Note: because TransactionIds are 32 bits and wrap around at 0xFFFFFFFF,
+ * CSNLog page numbering also wraps around at
+ * 0xFFFFFFFF/CSN_LOG_XACTS_PER_PAGE, and CSNLog segment numbering at
+ * 0xFFFFFFFF/CLOG_XACTS_PER_PAGE/SLRU_PAGES_PER_SEGMENT.  We need take no
+ * explicit notice of that fact in this module, except when comparing segment
+ * and page numbers in TruncateCSNLog (see CSNLogPagePrecedes).
+ */
+
+/* We store the commit CSN for each xid */
+#define CSN_LOG_XACTS_PER_PAGE (BLCKSZ / sizeof(XidCSN))
+
+#define TransactionIdToPage(xid)	((xid) / (TransactionId) CSN_LOG_XACTS_PER_PAGE)
+#define TransactionIdToPgIndex(xid) ((xid) % (TransactionId) CSN_LOG_XACTS_PER_PAGE)
+
+/*
+ * Link to shared-memory data structures for CLOG control
+ */
+static SlruCtlData CSNLogCtlData;
+#define CsnlogCtl (&CSNLogCtlData)
+
+static int	ZeroCSNLogPage(int pageno, bool write_xlog);
+static void ZeroTruncateCSNLogPage(int pageno, bool write_xlog);
+static bool CSNLogPagePrecedes(int page1, int page2);
+static void CSNLogSetPageStatus(TransactionId xid, int nsubxids,
+									  TransactionId *subxids,
+									  XidCSN csn, int pageno);
+static void CSNLogSetCSNInSlot(TransactionId xid, XidCSN csn,
+									  int slotno);
+
+static void WriteXidCsnXlogRec(TransactionId xid, int nsubxids,
+					 TransactionId *subxids, XidCSN csn);
+static void WriteZeroCSNPageXlogRec(int pageno);
+static void WriteTruncateCSNXlogRec(int pageno);
+
+/*
+ * CSNLogSetCSN
+ *
+ * Record XidCSN of transaction and its subtransaction tree.
+ *
+ * xid is a single xid to set status for. This will typically be the top level
+ * transactionid for a top level commit or abort. It can also be a
+ * subtransaction when we record transaction aborts.
+ *
+ * subxids is an array of xids of length nsubxids, representing subtransactions
+ * in the tree of xid. In various cases nsubxids may be zero.
+ *
+ * csn is the commit sequence number of the transaction. It should be
+ * AbortedCSN for abort cases.
+ */
+void
+CSNLogSetCSN(TransactionId xid, int nsubxids,
+					 TransactionId *subxids, XidCSN csn, bool write_xlog)
+{
+	int			pageno;
+	int			i = 0;
+	int			offset = 0;
+
+	Assert(TransactionIdIsValid(xid));
+
+	pageno = TransactionIdToPage(xid);		/* get page of parent */
+
+	// TODO(pooja): Remove this part entirely?
+	// if(write_xlog)
+	// 	WriteXidCsnXlogRec(xid, nsubxids, subxids, csn);
```

Comment:
So there are 2 type of WAL records: XLOG_CSN_ASSIGNMENT and XLOG_CSN_SETXIDCSN.
XLOG_CSN_ASSIGNMENT records the assignment happening in precommit and XLOG_CSN_SETXIDCSN records the mapping.

However, it turns out that the assigned CSN is derived from XactLastRecEnd` which is the last LSN of the transaction (and may not necessary be the commit record?). 

So when we reconstruct the csn mapping, we ideally need to look at XLOG_CSN_ASSIGNMENT, but this record only contains the CSN so I think we still need the XLOG_CSN_SETXIDCSN record and this code should be uncommented.

I may miss it, but it is unclear to me what would happen if postgres crashes after the last LSN is flushed to disk but before either XLOG_CSN_ASSIGNMENT or XLOG_CSN_SETXIDCSN is flushed to disk. However, this doesn't affect the experiment so let's ignore it for now.

Reply:
I think this depends on how we handle it in the pageserver upon recovery. If we consider XactLastRecEnd as the marker for CSN, then we don't need to care about either  XLOG_CSN_ASSIGNMENT or XLOG_CSN_SETXIDCSN records. I was thinking of handling it this way in the page server. Does that make sense?

Reply:
Did you mean we look at the last WAL record of a transaction to determine its CSN? I wasn't sure if that was possible since I didn't know if we could tie a log record to a transaction, but I just found that each WAL record had the `xl_xid` field that stored the associated transaction id, which could make things easier.

However, we still need to double-check that the CSN is the LSN of the very last log record. One case I can think of is when parallel workers of the same transaction append to the WAL at the same time and the last LSN of one worker might not be the last LSN of the transaction. The code might already handle this, so we need to confirm that. 

In any case, we can fallback to using the XLOG_CSN_SETXIDCSN record.